### PR TITLE
Limelight Vision

### PR DESCRIPTION
**Pull request by Mboesn**

### CODE REVIEW

File: src/main/java/frc/robot/pidsources/VisionPIDSourceY.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    public void setPIDSourceType(PIDSourceType pidSource) {
+    }
+
+    @Override
+    public double pidGet() {
+        return Robot.limelight.getTy();
+    }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/pidsources/VisionPIDSourceY.java:24]: Robot.limelight.getTy(); to .getDistance();

File: src/main/java/frc/robot/utils/Limelight.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /**
+     * @return The distance between the the target and the limelight
+     */
+    //TODO: set real function
+    public double getDistance() {
+        double x = getTy();
+        return -114.6 * Math.log(x) - 111.94;
+    }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Limelight.java:75]: add TODO set the log distance function

File: src/main/java/frc/robot/pidsources/VisionPIDSourceX.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    public void setPIDSourceType(PIDSourceType pidSource) {
+    }
+
+    @Override
+    public double pidGet() {
+        return Robot.limelight.getTx();
+    }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/pidsources/VisionPIDSourceX.java:24]: Robot.limelight.getTx(); to .getAngle()

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
change now elses things and then the command with me

File: src/main/java/frc/robot/utils/Limelight.java (hunk 2 above)
Comment (Mboesn) [reply src/main/java/frc/robot/utils/Limelight.java:75]: I did

File: src/main/java/frc/robot/commands/VisionPID.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ * A command for turning in place and driving straight with vision correction.
+ */
+public class VisionPID extends Command {
+  PIDController pidControllerX, pidControllerY;
+  PidSettings pidSettingsX, pidSettingsY;
+  Button button;
+  Target target;
+  double timeOnTargetX, timeOnTargetY, yOutput, xOutput;
+  Supplier<Double> forwardSupplier = () -> 0.0;
+  boolean isPIDOnY, onTarget;
```
Comment (Mboesn) [inline src/main/java/frc/robot/commands/VisionPID.java]: change isPIDOnY to a better name like isDrivingForward

File: src/main/java/frc/robot/commands/VisionPID.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  /** The robot will drive straight with the correction of the vision. */
+  public VisionPID(Target target, Supplier<Double> forwardSupplier, Button button) {
+    requires(Robot.drivetrain);
+    this.target = target;
+    this.forwardSupplier = forwardSupplier;
+    this.pidSettingsX = RobotConstants.RobotPIDSettings.VISION_X_PID_SETTINGS;
+    this.button = button;
+    this.isPIDOnY = false;
```
Comment (Mboesn) [inline src/main/java/frc/robot/commands/VisionPID.java]: unneccery

File: src/main/java/frc/robot/commands/VisionPID.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    if (!pidControllerX.onTarget() || Robot.limelight.getTv())
+      timeOnTargetX = Timer.getFPGATimestamp();
+
+    if (!pidControllerY.onTarget() || Robot.limelight.getTv())
+      timeOnTargetY = Timer.getFPGATimestamp();
+  }
+
+  @Override
+  protected boolean isFinished() {
+    if (isPIDOnY)
```
Comment (Mboesn) [inline src/main/java/frc/robot/commands/VisionPID.java]: add docs

**Review by Mboesn — COMMENTED** [review]
also delete files

**Pull request merged**